_posts/2020-07-21-cube-functionality-to-kartothek.markdown

```diff
@@ -0,0 +1,171 @@
+>>> dfs = query_cube(
+...     cube=cube,
+...     store=store,
+...     partition_by="country",
+... )
+>>> dfs[0]
+   avg_temp     city country        day   latitude  longitude
+0         8  Dresden      DE 2020-07-01  51.050407  13.737262
+1         4  Dresden      DE 2020-07-02  51.050407  13.737262
+2         6  Hamburg      DE 2020-07-01  53.551086   9.993682
+3         5  Hamburg      DE 2020-07-02  53.551086   9.993682
+>>> dfs[1]
+   avg_temp    city country        day   latitude  longitude
+0         6  London      UK 2020-07-01  51.509865  -0.118092
+1         8  London      UK 2020-07-02  51.509865  -0.118092
+```
+
+The query system also supports selection and projection:
+
+```python
+>>> from kartothek.core.cube.conditions import C
+>>> from kartothek.io.eager_cube import query_cube
+>>> query_cube(
+...     cube=cube,
+...     store=store,
+...     payload_columns=["avg_temp"],
+...     conditions=(
+...         (C("country") == "DE") &
+...         C("latitude").in_interval(50., 52.) &
+...         C("longitude").in_interval(13., 14.)
+...     ),
+... )[0]
+   avg_temp     city country        day
+0         8  Dresden      DE 2020-07-01
+1         4  Dresden      DE 2020-07-02
+```
+
+
+## Additional Features of Cube
+
+*	**Multi-dataset with Single Table**: When mapping multiple parts (tables or datasets) to Kartothek, using multiple datasets allows users to copy, backup and delete them separately. 
+Index structures are bound to datasets which feels more consistent.
+*	**Explicit physical Partitions**: We have decided for explicit physical partitions since we have seen that this data model works well for our current data flow. 
+It allows quick and efficient re-partitioning to allow row-based, group-based, and timeseries-based data processing steps, while keeping the technical complexity rather low (compared to an automatic + dynamic partitioning).
+It also maps well to multiple backends we planned to use.
+*	**Update Granularity Partition-wise**: Entire partitions can overwrite old physical partitions. Deletion operations are partition-wise.
+*	**Seed-Based Join System / Partition-alignment**: When data is stored in multiple parts (tables or datasets), the question is how to expose it to the user during read operations.
+ Seed based Join marks a single part as seed which provides seed dataset in the cube, all other parts are just additional columns.
+ Cube uses lazy approach of seed based join, 
+ since it better supports independent copies and backups of datasets and also simplifies some of our processing pipelines (e.g. geolocation data can blindly be fetched for too many locations and dates.)	
+
+## Dask And Yamal Support
+
+The following features are supported by different backends:
+
+
+| Feature | Eager | Dask.Bag | Dask.DataFrame |	Yamal |
+| :-------: | :-----: | :--------: | :--------------: | :-----: |
+| Build	  |    yes    |    yes     |      yes         |   yes   |
```

Comment:
Would be nice if the table would be readable in source code as well (aka being aligned)

Reply:
done.